### ISSUE TITLE
How to use onClick and onDrag together?

### ISSUE THREAD

**Issue by codypersinger** [react-spring/react-use-gesture#66]
I'm using react-use-gesture and react-spring to create moveable divs like so: https://codesandbox.io/s/xmqwz

```
import React, { useState } from 'react'
import { render } from 'react-dom'
import { useGesture } from 'react-use-gesture'
import { useSprings, animated } from 'react-spring'
import './styles.css'

const addVectors = (v1, v2) => [v1[0] + v2[0], v1[1] + v2[1]]

const Moveables = ({ items }) => {
  const [selectedItems, setSelectedItems] = useState([])

  const toggleSelectedItem = item => {
    const { id } = item
    if (selectedItems.some(item => item.id === id)) {
      setSelectedItems(selectedItems.filter(item => item.id !== id))
    } else {
      setSelectedItems(oldState => [...oldState, item])
    }
  }

  const [springs, setSprings] = useSprings(items.length, index => ({
    xy: items[index].position,
    zIndex: 0
  }))

  const binds = useGesture({
    onDrag: ({ args: [itemIndex], delta, temp = springs[itemIndex].xy.getValue() }) => {
      const selectedItemIndexes = []
      items.forEach(({ id }, index) => {
        if (selectedItems.some(item => item.id === id)) {
          selectedItemIndexes.push(index)
        }
      })

      setSprings(index => {
        const animationValues = selectedItemIndexes.includes(index)
          ? {
              xy: addVectors(delta, temp),
              zIndex: 1
            }
          : undefined
        return animationValues
      })
      return temp
    }
  })

  return (
    <div className="content">
      {springs.map(({ xy, zIndex }, index) => {
        const item = items[index]
        const { id } = item
        return (
          <animated.div
            key={id}
            style={{
              zIndex,
              transform: xy.interpolate((x, y) => `translate3d(${x}px, ${y}px, 0)`),
              border: selectedItems.some(item => item.id === id) ? '4px solid lightgreen' : '0'
            }}
            {...binds(index)}
            onDoubleClick={() => toggleSelectedItem(item)}>
            {id}
          </animated.div>
        )
      })}
    </div>
  )
}

render(
  <Moveables
    items={[
      {
        id: 0,
        position: [0, 100]
      },
      {
        id: 1,
        position: [0, 200]
      },
      {
        id: 2,
        position: [0, 0]
      }
    ]}
  />,
  document.getElementById('root')
)
```

When double-clicking on a div, it will become selected. When dragging, all selected divs should move at the same time.

This already works great - except after selecting one div so that it's draggable, double-clicking on any other div in an attempt to select it will cause the first already-selected div to move.

How can I set this up so that click events don't trigger the onDrag gesture?

**Comment by dbismut** [comment]
Hey, thanks for reporting the issue. Do you mind creating a functional (simplified?) codesandbox reproducing the problem?

**Comment by codypersinger** [comment]
No problem, here is the codesandbox showing the issue. Try double-clicking to select two divs and move them both at the same time. Selecting the second div is difficult: https://codesandbox.io/s/xmqwz

I've noticed that setting a different `top` value for each `.content > div` in the CSS and zeroing out the initial `xy` values will somewhat alleviate the issue, as you can see here: https://codesandbox.io/s/blazing-resonance-gw9pj

I suppose one solution is calculating the offset for each div based on the height (or size) - kind of like the draggable list example.

However, what I'd really like is some way to maintain the offsets of all the divs relative to each other. Imagine selecting multiple icons to drag/drop on the desktop, those icons keep their relative distance to one another while dragging. I think my issue is that the drag is trigger on click, so when clicking to select the second div, the first selected div is dragged towards the second one.

**Comment by dbismut** [comment]
Hey, thanks for putting up the codesandbox!
Is that what you want to achieve? https://codesandbox.io/s/awesome-surf-414f1

I'm pretty sure this can be further simplified, but let me know if this works as is?

**Comment by codypersinger** [comment]
Yes, that's *exactly* what I wanted, and I don't think I would've figured that out on my own! It looks like the main trick is mapping the temp value? I'm sure I'll use this much more in the future, thanks very much for taking the time to help and teach.

**Comment by dbismut** [comment]
Yes, that’s the main trick. Essentially, `useGesture` keeps track of just one state, which is relative to pointer coordinates. What you usually want to do is move objects based on the mouse movement (ie the `delta` value). But you need to know the origin position of each objects prior to the mouse movement so that you can add it to the `delta` value.

Note that in a simple case where you only have one object without needing to enable / disable drag, this is exactly what the `local` attribute is designed for.

**Comment by codypersinger** [comment]
@dbismut I've been playing around with the library more, and have come across a situation or two where it would still be useful to differentiate between a click and a drag event. Is there some way to do this, using the previous example in this thread? Thank you for your time!

**Comment by dbismut** [comment]
That’s a bit trickier, as drag gesture is currently triggered by `onMouseDown`, so at the moment a click event would inevitably trigger a drag gesture. One way to solve this would be to only trigger a drag event when the pointer actually moves or after a short delay. I’ll see if I can do this without generating regressions.

**Comment by codypersinger** [comment]
I thought that might be the case - and I had the same thought about triggering a click only if the pointer has moved, but I thought that might tend to trigger false clicks if someone was intending to drag but just hadn't decided where to move yet. Even with the delay, it doesn't sound like a concretely reliable solution, but it would likely be better than nothing.

I was wondering if it might be possible to initiate dragging on DragStart or Drag event, but I know from having implemented custom dnd solutions in the past that there are a lot of caveats and knock-on effects to consider. I need to take a closer look at the useGesture source and get a better understanding of what would be possible.

**Comment by dbismut** [comment]
Just to clarify the idea would be to trigger a drag to the earliest condition:
- after a short delay following a mousedown event
- whenever the pointer moves after a mousedown

I think this is the logic adopted by [hammerjs](https://hammerjs.github.io/). Unfortunately I’m not sure there is a perfect alternative.

**Comment by codypersinger** [comment]
I agree, there is likely no perfect solution. I think what you've proposed is probably as good as it gets, and I think it would work well enough in the vast majority of situations.

**Comment by dbismut** [comment]
onDragStart would require to make the element draggable right (ie attribute `draggable` set to `true`)? That means conflicting with the native draggable api, we’ve tried that before and it’s a tough battle.

**Comment by codypersinger** [comment]
Yes, that's right. The native api absolutely leaves a lot to be desired. I've created one implementation in the past that goes this route, within React, and worked on the web as well as across all native devices that we tested with (at a big agency, so pretty extensive iOS/Android testing) - but it absolutely involved fighting against the native implementation. I might play around with this idea in a different branch. The implementation would be pretty extensive (probably too extensive to be merged back, perhaps a fork would be better), and there may be issues I'm overlooking. I think your initial suggestion is much more immediately useful and workable.

**Comment by dbismut** [comment]
Hey @codypersinger I have something working where drag only happens after a short delay or the mouse moving, but I just have one question: what's the difference between a drag and a click in your use case? The question is a bit silly, but is it the fact that the pointer should not move during a click, or is it that the click is supposed to be short?

**Comment by codypersinger** [comment]
Awesome! I'm still playing around with the native API implementation, making slow progress. It's not a silly question at all. For pretty much all the use cases I can think of, the click being short is the more important and reliable factor.

**Comment by dbismut** [comment]
Maybe there's no need for what I was mentioning above then:
https://codesandbox.io/s/awesome-surf-414f1

It's a bit silly, but I think this is more or less what you want isn't it?

**Comment by codypersinger** [comment]
It does more or less achieve the effect. It's fairly easy to both drag and click with quick gestures, which is not ideal behavior in the example codepen - I think making sure the pointer hasn't moved would take care of that. It would be nice to have an `onClick` binding abstraction which is configurable with props like `duration` and `allowMovement`, so that click-based logic doesn't have to live in "onDrag" - but I understand that those events are necessarily intertwined, and it may not jive with the goals of this library which seem drag-focused. What are your thoughts?

**Comment by dbismut** [comment]
I've just published it on npm with the `next` tag. `yarn add react-use-gesture@next` should do it.

**Comment by codypersinger** [comment]
I'm not noticing any difference in when `react-use-gesture`'s `onDrag` is triggering in regards to either delay or cursor position after upgrading to `5.2.0-beta.1`, but perhaps I'm doing something wrong? I have to run to a meeting now but I'll knock together a codepen ASAP

**Comment by dbismut** [comment]
I haven’t tested it myself except locally tbh so maybe there’s something wrong!

**Comment by codypersinger** [comment]
I'd be happy to take a look sometime tonight, if there's somewhere I can see the commits / history for that release? Thanks for all your time and effort on this so far, I truly appreciate it.

**Comment by dbismut** [comment]
Have left my computer at work :) will push the branch tomorrow morning CET time.

**Comment by codypersinger** [comment]
Smart choice! That sounds great, no rush. :v:

**Comment by dbismut** [comment]
Hey, so I just pushed the code on the `next` branch (tests are failing but that's expected).

Have a look at: https://codesandbox.io/s/dazzling-vaughan-4j8dd
I've set the time threshold to 1 second so that the lag is noticeable.

1. If you just briefly click on the pink square, only the click counter should increment.
2. If you mouse down for 1 second or more **without moving your mouse**, the drag status should show _drag started_. 
3. If you release your mouse while hovering the pink square, the  the drag status should show _drag ended_ and the click counter should increment
4. If you mouse down and then move your mouse right after it should trigger a drag right away, without waiting for 1 second

All of this is expected in the way I've implemented the logic. Note that the click handler is untouched, so it behaves exactly like it should.

PS: after looking at Hammer JS code, they actually trigger a drag only after the pointer moves past a threshold (default is `10px`).

**Comment by codypersinger** [comment]
Very nice, this works extremely well. The only issue I've noticed is with `immeditate: true` when calling `set`, the `click` counter is incremented when the drag ends. Other than that, this looks flawless from my admittedly limited testing.

**Comment by dbismut** [comment]
Well yes... as I was saying, if you release your mouse over the dragged element it’s still considered as a click by the browser. `immediate:true` pretty much guarantees that the dragged element follows your mouse so that’s why the click counter increments. Theres not much we can do about this.

**Comment by codypersinger** [comment]
I see what you mean, my mistake.

**Comment by codypersinger** [comment]
Is it possible to optionally prevent the click from being triggered if you are actively dragging and release the mouse over the pink square?

**Comment by dbismut** [comment]
It is, but not natively in the lib. You can use the same trick from my sandbox above:

https://codesandbox.io/s/awesome-surf-414f1

**Comment by codypersinger** [comment]
That makes sense. Here's a simple example for anyone that sees this in the future: https://codesandbox.io/s/exciting-surf-89ncy

**Comment by dmcmorris** [comment]
Looks like this never made it out of beta... anyone have a solution to this issue?  Trying to use onDrag from react-use-gesture and onDblClick from konvajs and I always get a drag triggered for every double click

**Comment by dbismut** [comment]
Hey @dmcmorris, indeed drag is triggered as soon as the user mouses down. Unless proven otherwise, I’ve found difficult to accurately predict whenever a drag should actually be triggered, this depends on the usecase. 

It shouldn’t be a big deal in your case though, nothing forces you to update your UI when the drag handler triggers. A double click generally happens with no mouse movement, so you could only update your UI in `onDrag` when `first` or `last` are `false`. If you actually need something specific to happen when the user starts or ends dragging, you could use a ref. I can explain this in greater details if you provide me with a concrete use case.

**Comment by codynova** [comment]
@dmcmorris The solution can be seen [here ](https://codesandbox.io/s/exciting-surf-89ncy) - you need to use `first` and `last` in `useGesture`'s `onDrag`, and store the drag delta time in a ref.

`onDrag`:
```
	if (first && time) {
		// Drag started, no click triggered
		dragDeltaTime.current = time;
	}
	else if (last && time) {
		// Drag ended, no click triggered
		dragDeltaTime.current = time - dragDeltaTime.current;
	}
```

`onClick`:
```
	event.stopPropagation();

	if (dragDeltaTime.current < 60) {
		// Click started, no drag triggered
	}

	dragDeltaTime.current = 0;
```

**Comment by dmcmorris** [comment]
thanks @codynova  and @dbismut, I had not fully explored all of the attributes available in onDrag.  For my purposes I used distance to filter/control the UI update and ignore drags:

onDrag:
```
        if (distance < minDragDist) {  //minDragDist = 10
          return memo;
        }
```

The quick replies are much appreciated... awesomely powerful library.

**Comment by dbismut** [comment]
Just be careful with distance as dragging might start but then if you get back to your point of origin without releasing your drag, distance might pass under the threshold and your UI will not be updated.

**Comment by codynova** [comment]
>Unless proven otherwise, I’ve found difficult to accurately predict whenever a drag should actually be triggered, this depends on the use case.

@dbismut For the record, using the click speed/delta time solution has been extremely accurate in my experience - although perhaps it depends on the use case, as you said. However this has been my go-to solution for complex UIs that strongly differentiate between clicking and dragging on the same element and it has always seemed perfectly accurate in my experience, I have never seen a false-positive.

**Comment by dmcmorris** [comment]
I'm using a ref to track if the drag has passed the distance threshold or not, I need that to filter other events/logic anyways.  I have some momentum scrolling using spring on this element and longpress events, so time based initiation of the drag was more problematic for me:

```
const dragInfo = React.useRef({ isDragging: false });

...

onDrag: ({
        active,
        movement,
        velocity,
        direction,
        distance,
        first,
        last,
        memo = position.getValue()
      }) => {

...
        if (!dragInfo.current.isDragging && distance < minDragDist) {
          return memo;
        }

        dragInfo.current.isDragging = true;

        if (last) {
          dragInfo.current.isDragging = false;
        }
...

       return memo;
  }
);
```

I'll try to throw a simple sandbox together and share later.  Thanks again for the concise and helpful suggestions.

**Comment by dbismut** [comment]
@codynova that’s good to know and I think I’ll add your sandbox to examples / FAQ. However native click events don’t depend on time or movement (I think they’re triggered as long as the target is still under the mouse when it’s released).

**Comment by dbismut** [comment]
@dmcmorris that’s perfect indeed. No need for a sandbox ;)

**Comment by codynova** [comment]
@dbismut I made a second sandbox that's a bit clearer, and shows drag/click/double-click all working together. https://codesandbox.io/s/elated-cori-45z3q

What do you mean exactly in regards to native events? Can you describe the effect that would have on the sandbox?

**Comment by dmcmorris** [comment]
@dbismut one thing to point out, the `{time: 1000}` param doesn't seem to do anything with any versions of react-use-gesture but 5.0.2_beta1

**Comment by codynova** [comment]
@dmcmorris Good point. I'm not sure whether that is will actually make it into the library at some point, but it is not necessary for the solution.

**Comment by dbismut** [comment]
You’d just have to ask, I don’t have anything against it 🙃 it would be set to 0 by default.

**Comment by codynova** [comment]
Sorry if that sounded rude! I realized my comment can be read two different ways - and I definitely didn't intend to throw shade 😅

Honestly I don't have an immediate use case for it, and I don't want to suggest adding a lot of clutter to the API - but I do feel like it could be useful, and I appreciate having access to parameters that allow edge-case solutions.

As far as a name, assuming it effects all gesture types, maybe something like `gestureDelay`?

**Comment by dbismut** [comment]
Oh no don’t worry, I was joking! I’m not sure what other gestures would make sense to have that feature, so for now I think I’ll just make it work for drag (`dragDelay` it should be).

It’s not that much of an edge case, being able to distinguish a click from a drag should be pretty common and if this avoids hacks or refs its a good thing to have!

**Comment by dbismut** [comment]
There, `v6.0.4` is out with `dragDelay` as an option (set to `false` by default, you can set it to `true` and will take a default value of `180ms` but you can set any value you want).

**Comment by codynova** [comment]
Awesome! I've updated that [sandbox](https://codesandbox.io/s/xenodochial-dubinsky-l9x7f). That demo still uses refs to determine click and double-click, and uses `dragDelay` to allow initiating a drag event without moving the mouse (by holding in click for `dragDelay` milliseconds).

I suppose it would also be nice to have an example that only uses `dragDelay` without refs. I'm a bit busy right this second but can try to throw something together soon

**Comment by codynova** [comment]
@dbismut Just doing some testing and there seems to be a significant difference in behavior between 5.2.0-beta.1 `time` and 6.0.4 `dragDelay`? Here are some example codepens to illustrate the difference:

[5.2.0-beta.1 no time config](https://codesandbox.io/s/sharp-mirzakhani-zmsbw). This setup triggers a drag event very soon after the mouse click is held.

[5.2.0-beta.1 with time config](https://codesandbox.io/s/xenodochial-dubinsky-l9x7f). This setup triggers a drag event after the mouse click is held for `time` milliseconds. In this case, 1000ms.

[6.0.4 with dragDelay config](https://codesandbox.io/s/laughing-liskov-62st6). I would expect with to behave the same as 5.2.0-beta.1 with time config, but dragging no longer seems to work properly.

**Comment by codynova** [comment]
@dbismut Please disregard my last comment. My example sandbox breaks when updated to 6.0.4, and I initially thought it was due to the `dragDelay` setting, but now I can see that's not the problem. Here's the [original sandbox](https://codesandbox.io/s/stoic-kilby-e4vnr), and that same [sandbox updated to 6.0.4](https://codesandbox.io/s/crazy-noyce-9l1td). Can you help me understand the difference in the drag behavior between the two?

**Comment by dbismut** [comment]
@codynova yep that's an easy one, v6 labels things in a better / more conventional way.

Value | v5.x | v6.x
------------ | -------------|---------
delta between current and previous event | N/A | **delta**
last gesture offset (xy - initial) | delta | **movement**
offset since first gesture | local | **offset**

So you just need to replace `delta` with `movement` and you should be good ;)

**Comment by codynova** [comment]
@dbismut Thanks very much for the speedy reply, I've made the change as suggested - but things still don't seem to be working as I'd expect.

[5.2.0-beta.1 sandbox](https://codesandbox.io/s/elastic-rosalind-6jfpu) with `time` configured to 1000ms (using `delta` instead of `movement`). With this setup, you can click, double-click, and click-and-drag without delay. You can also click-and-hold for 1000ms to initiate a drag.

[6.0.4 sandbox](https://codesandbox.io/s/youthful-hooks-j3915) with `dragDelay` configured to 1000ms (using `movement` instead of `delta`). With this setup, you can click and double-click, but click-and-drag is not working properly. You can, however, still click-and-hold for 1000ms to initiate a drag, which works as I'd expect.

**Comment by dbismut** [comment]
Yeah in 6.0.4 I haven't made so that the drag is initiated as soon as the mouse moves (it really waits for the 1000ms). I can fix this in 6.0.5, I'll see what I can do now :)

**Comment by codynova** [comment]
Awesome. Hopefully the update in 6.0.5 won't make `dragDelay` confusing for consumers of the API. I suppose it might be possible to have a second parameter `instantDragOnMove` or `dragDelayOnMove` or something, which would allow utilizing `dragDelay` in either the 6.0.4 or 6.0.5 fashion. Honestly without foreseeing exact use-cases, I think it's kind of hard to say whether it would be useful. Thank you very much for the great and speedy work on this lib!

**Comment by dbismut** [comment]
Yeah `dragDelay` is optin and pretty sure people will stumble upon this thread if they have any questions ;) 6.0.5 is now out, let me know if it works better for you!

**Comment by codynova** [comment]
Love it. https://codesandbox.io/s/trusting-pine-jpqi5

I cleaned things up a bunch, and I think this is the best solution I can come up with. I'll be replacing the logic in my current projects with the logic from this sandbox. It requires three different refs, but it's working extremely well in my testing.

Rather than tracking drag delta time, we only use a timestamp ref for tracking the time of the last click, and we use boolean refs for tracking current drag and double-click status. And instead of using `onClick` and `onDoubleClick` both (and duplicating click delta time logic), we put the double click logic inside the `onClick` handler.

I tried simply using the `dragStatus` value to eliminate the need for the two boolean refs, but it was not nearly as accurate/performant as using the refs.

With this setup, it is (intentionally) not possible to trigger two double-clicks in a row. Clicking fast enough to be within the `DOUBLE_CLICK_TIME_THRESHOLD` will simply trigger a regular click, in the previous click was a double-click. If one would like to allow triggering simultaneous double-clicks, remove the `isDoubleClicked` ref and relevant logic.

**Comment by dbismut** [comment]
Hey @codynova I've found a bug with `dragDelay` this morning: in `6.0.6` if you moused down, moved and dragged longer than dragDelay, the square would go back to its origin position (I had forgotten to clear a timeout). This is now fixed in `6.0.7`.
I've copied your sandbox here and made a few tweaks, mainly moving to react-spring v9 which simplifies styling attributes: https://codesandbox.io/s/hardcore-kowalevski-kepmr

Also I think you can use `performance.now()` without using `window`, it should be better in case you ever move to SSR where `window` is undefined.

**Comment by SidKH** [comment]
For anyone looking to implement `dragDelay` option from this thread, it has been renamed in release `7.0.0`

https://github.com/react-spring/react-use-gesture/blob/fa6dd05957c34450533445224a42fdc89852be68/CHANGELOG.md#breaking-changes

**Comment by Sorkstryparen** [comment]
Is there any example with v.10 regarding this?
Can't get @dbismut sandbox example above to work with latest versions of react-gesture and react-spring.
Even after changing dragDelay to delay 🤔
